File: hosts/macos/Sources/EngawaHost/ProcessAdapter.swift
import Foundation
import EngawaKit

// The `process` namespace (spec/commands/process.md). Pull streams (§4.1): the host buffers
// per stream, signals process.readable when a drained buffer goes non-empty, and drains via
// process.read; process.exit fires only once both streams are drained. 8 MiB per-stream cap
// applies OS backpressure by pausing pipe reads. Sidecars are allowlisted per §7.2.
final class ProcessAdapter: Adapter, @unchecked Sendable {
    let namespace = "process"

    private let manifest: Manifest?
    private let emitter: EventEmitter
    private let lock = NSLock()
    private var procs: [Int: Proc] = [:]
    private let bufferCap = 8 * 1024 * 1024

    final class Stream {
        let handle: FileHandle
        var buffer = Data()
        var eof = false
        var paused = false
        init(_ handle: FileHandle) { self.handle = handle }
    }

    final class Proc {
        let process: Process
        let stdin: FileHandle
        let stdout: Stream
        let stderr: Stream
        var exited = false
        var exitCode = 0
        var exitEmitted = false
        init(process: Process, stdin: FileHandle, stdout: FileHandle, stderr: FileHandle) {
            self.process = process
            self.stdin = stdin
            self.stdout = Stream(stdout)
            self.stderr = Stream(stderr)
        }
    }

    init(manifest: Manifest?, emitter: EventEmitter) {
        self.manifest = manifest
        self.emitter = emitter
    }

    func handle(_ cmd: String, _ args: JSONValue) async throws -> JSONValue {
        switch cmd {
        case "spawn": return try spawn(args)
        case "stdinWrite": return try stdinWrite(args)
        case "read": return try read(args)
        case "kill": return try kill(args)
        default: throw AdapterError("ENOSYS", "unknown command: process.\(cmd)")
        }
    }

    // MARK: spawn

    private func spawn(_ args: JSONValue) throws -> JSONValue {
        let obj = args.objectValue ?? [:]
        guard let command = obj["command"]?.stringValue, !command.isEmpty else {
            throw AdapterError("EINVAL", "command required")
        }
        guard let manifest = manifest, let exePath = manifest.resolveSidecar(command) else {
            throw AdapterError("EPERM", "not a declared in-bundle sidecar: \(command)")
        }
        let cmdArgs = (obj["args"]?.arrayValue ?? []).compactMap { $0.stringValue }

        let process = Process()
        process.executableURL = URL(fileURLWithPath: exePath)
        process.arguments = cmdArgs
        let stdinPipe = Pipe(), stdoutPipe = Pipe(), stderrPipe = Pipe()
        process.standardInput = stdinPipe
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        do {
            try process.run()
        } catch {
            throw AdapterError("EIO", "failed to spawn: \(error.localizedDescription)")
        }

        let pid = Int(process.processIdentifier)
        let proc = Proc(process: process,
                        stdin: stdinPipe.fileHandleForWriting,
                        stdout: stdoutPipe.fileHandleForReading,
                        stderr: stderrPipe.fileHandleForReading)

        lock.lock(); procs[pid] = proc; lock.unlock()

        proc.stdout.handle.readabilityHandler = { [weak self] fh in self?.onData(pid: pid, which: "stdout", fh: fh) }
        proc.stderr.handle.readabilityHandler = { [weak self] fh in self?.onData(pid: pid, which: "stderr", fh: fh) }
        // Look the process up by pid rather than capturing the non-Sendable Proc.
        process.terminationHandler = { [weak self] p in
            self?.onTermination(pid: pid, status: Int(p.terminationStatus))
        }

        return .object(["pid": .number(Double(pid))])
    }

    private func onTermination(pid: Int, status: Int) {
        lock.lock()
        if let proc = procs[pid] { proc.exited = true; proc.exitCode = status }
        lock.unlock()
        maybeEmitExit(pid: pid)
    }

    // MARK: stream ingest

    private func onData(pid: Int, which: String, fh: FileHandle) {
        let data = fh.availableData

        lock.lock()
        guard let proc = procs[pid] else { lock.unlock(); return }
        let stream = which == "stdout" ? proc.stdout : proc.stderr

        if data.isEmpty {
            stream.eof = true
            fh.readabilityHandler = nil
            lock.unlock()
            maybeEmitExit(pid: pid)
            return
        }

        let wasEmpty = stream.buffer.isEmpty
        stream.buffer.append(data)
        if stream.buffer.count >= bufferCap {
            stream.paused = true
            fh.readabilityHandler = nil   // stop draining the pipe → OS backpressure
        }
        lock.unlock()

        if wasEmpty {
            emitter.emit("process.readable", .object(["pid": .number(Double(pid)), "stream": .string(which)]))
        }
    }

    // MARK: read

    private func read(_ args: JSONValue) throws -> JSONValue {
        let obj = args.objectValue ?? [:]
        guard let pidD = obj["pid"]?.numberValue else { throw AdapterError("EINVAL", "pid required") }
        let pid = Int(pidD)
        let which = obj["stream"]?.stringValue ?? "stdout"
        let maxBytes = Int(obj["maxBytes"]?.numberValue ?? 65536)

        lock.lock()
        guard let proc = procs[pid] else { lock.unlock(); throw AdapterError("ESRCH", "no such process: \(pid)") }
        let stream = which == "stdout" ? proc.stdout : proc.stderr

        let take = min(max(maxBytes, 0), stream.buffer.count)
        let (text, consumed) = Self.validUTF8Prefix(stream.buffer, upTo: take)
        if consumed > 0 { stream.buffer.removeFirst(consumed) }

        // Resume the pipe if it was paused for backpressure and has fallen below the cap.
        if stream.paused && !stream.eof && stream.buffer.count < bufferCap {
            stream.paused = false
            stream.handle.readabilityHandler = { [weak self] fh in self?.onData(pid: pid, which: which, fh: fh) }
        }
        let eof = stream.eof && stream.buffer.isEmpty
        lock.unlock()

        maybeEmitExit(pid: pid)
        return .object(["data": .string(text), "eof": .bool(eof)])
    }

    // MARK: stdinWrite / kill

    private func stdinWrite(_ args: JSONValue) throws -> JSONValue {
        let obj = args.objectValue ?? [:]
        guard let pidD = obj["pid"]?.numberValue else { throw AdapterError("EINVAL", "pid required") }
        guard let data = obj["data"]?.stringValue else { throw AdapterError("EINVAL", "data required") }
        lock.lock()
        let proc = procs[Int(pidD)]
        lock.unlock()
        guard let proc = proc else { throw AdapterError("ESRCH", "no such process: \(Int(pidD))") }
        do { try proc.stdin.write(contentsOf: Data(data.utf8)) }
        catch { throw AdapterError("EIO", "stdin write failed: \(error.localizedDescription)") }
        return .null
    }

    private func kill(_ args: JSONValue) throws -> JSONValue {
        let obj = args.objectValue ?? [:]
        guard let pidD = obj["pid"]?.numberValue else { throw AdapterError("EINVAL", "pid required") }
        lock.lock()
        let proc = procs[Int(pidD)]
        lock.unlock()
        guard let proc = proc else { throw AdapterError("ESRCH", "no such process: \(Int(pidD))") }
        proc.process.terminate()
        return .null
    }

    // MARK: exit

    private func maybeEmitExit(pid: Int) {
        lock.lock()
        guard let proc = procs[pid] else { lock.unlock(); return }
        let drained = proc.exited && !proc.exitEmitted
            && proc.stdout.eof && proc.stderr.eof
            && proc.stdout.buffer.isEmpty && proc.stderr.buffer.isEmpty
        if drained { proc.exitEmitted = true }
        let code = proc.exitCode
        lock.unlock()

        if drained {
            emitter.emit("process.exit", .object(["pid": .number(Double(pid)), "code": .number(Double(code))]))
        }
    }

    // Longest valid-UTF-8 prefix of the first `upTo` bytes, so a read never splits a codepoint.
    private static func validUTF8Prefix(_ data: Data, upTo: Int) -> (String, Int) {
        if upTo <= 0 { return ("", 0) }
        var end = upTo
        while end > 0 {
            let slice = data.prefix(end)
            if let s = String(data: slice, encoding: .utf8) { return (s, end) }
            end -= 1   // back off over an incomplete trailing sequence (≤3 bytes)
        }
        return ("", 0)
    }
}

File: hosts/windows/src/adapters/ProcessAdapter.cpp
// process namespace (spec/commands/process.md). Streams are PULL: process.readable signals data
// availability (no bytes ride the event, §4.1); JS drains with process.read. The host buffers up to
// 8 MiB per stream then applies OS backpressure by pausing the pipe read. process.exit fires only
// after both streams are drained. Only manifest-declared, in-bundle sidecars run (§7.2).
#include <windows.h>

#include <atomic>
#include <condition_variable>
#include <memory>
#include <mutex>
#include <thread>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "PathUtil.hpp"
#include "Utf.hpp"
#include "adapters/Adapters.hpp"

namespace {

constexpr size_t kCap = 8 * 1024 * 1024;
const char* kExeExts[] = {".exe", ".com", ".bat", ".cmd"};

struct StreamBuf {
    std::mutex gate;
    std::condition_variable resume;
    std::string data;         // backing store; the readable region is [off, data.size())
    size_t off = 0;           // read head — advancing it instead of erase(0,n) keeps draining O(n)
    bool eof = false;
    bool paused = false;
    bool everPaused = false;  // sticky: did backpressure ever pause this stream (__wasPaused)
    HANDLE readEnd = nullptr;

    size_t size() const { return data.size() - off; }  // bytes available to read
    bool empty() const { return off >= data.size(); }
    const char* head() const { return data.data() + off; }
    void append(const char* p, size_t n) { data.append(p, n); }
    // Advance past consumed bytes; compact only occasionally so a full drain is amortized O(n), not
    // the O(n^2) that erase(0,n) on every small read would cost against the 8 MiB cap.
    void consume(size_t n) {
        off += n;
        if (off >= data.size()) { data.clear(); off = 0; }
        else if (off >= 65536 && off * 2 >= data.size()) { data.erase(0, off); off = 0; }
    }
};

struct ProcState {
    HANDLE hProcess = nullptr;
    HANDLE job = nullptr;  // whole-process-tree kill (a sidecar may itself spawn children)
    DWORD pid = 0;
    HANDLE stdinWrite = nullptr;
    StreamBuf out;
    StreamBuf err;
    std::atomic<bool> exited{false};
    std::atomic<bool> exitEmitted{false};
    int exitCode = 0;
};

// Longest byte length ≤ take whose bytes form only COMPLETE UTF-8 sequences — so a read never emits a
// partial character, including when the tail is an incomplete multibyte sequence still awaiting
// continuation bytes on the pipe (mirrors the mock host).
int validUtf8End(const char* buf, int take) {
    int end = take < 0 ? 0 : take;  // caller passes take <= available bytes
    while (end > 0) {
        int i = end - 1;
        while (i > 0 && ((unsigned char)buf[i] & 0xc0) == 0x80 && end - i < 4) i--;
        unsigned char lead = buf[i];
        int seqLen;
        if ((lead & 0x80) == 0) seqLen = 1;
        else if ((lead & 0xe0) == 0xc0) seqLen = 2;
        else if ((lead & 0xf0) == 0xe0) seqLen = 3;
        else if ((lead & 0xf8) == 0xf0) seqLen = 4;
        else seqLen = 1;
        if (i + seqLen <= end) break;
        end = i;
    }
    return end;
}

std::unordered_set<std::string> loadSidecars(const std::string& bundleRoot) {
    std::unordered_set<std::string> set;
    auto o = Files::readJsonObject(U8(P(bundleRoot) / L"engawa.json"));
    if (o && o->contains("sidecars") && (*o)["sidecars"].is_array())
        for (auto& s : (*o)["sidecars"])
            if (s.is_string()) set.insert(s.get<std::string>());
    return set;
}

class ProcessAdapter : public IAdapter {
public:
    ProcessAdapter(IEventEmitter* emitter, const HostOptions& opts)
        : emitter_(emitter), conformance_(opts.conformance) {
        std::error_code ec;
        bundleRoot_ = U8(fsys::absolute(P(opts.bundleRoot), ec));
        sidecars_ = loadSidecars(bundleRoot_);
    }

    std::string ns() const override { return "process"; }

    Json handle(const std::string& command, const Json& args) override {
        if (command == "spawn") return spawn(args);
        if (command == "stdinWrite") return stdinWrite(args);
        if (command == "stdinClose") return stdinClose(args);
        if (command == "read") return read(args);
        if (command == "kill") return kill(args);
        if (command == "__wasPaused" && conformance_) return wasPaused(args);
        throw EngawaError::nosys("process." + command);
    }

private:
    // ---- spawn (§7.2 allowlist + platform executable resolution) --------------------------
    Json spawn(const Json& args) {
        std::string command = ja::reqString(args, "command");
        bool isBatch = false;
        std::string exe = resolveSidecar(command, isBatch);
        if (exe.empty()) throw EngawaError("EPERM", "not a declared in-bundle sidecar: " + command);

        std::vector<std::string> tokens;
        if (isBatch) {
            tokens.push_back(EnvOpt(L"ComSpec").value_or("cmd.exe"));
            tokens.push_back("/c");
            tokens.push_back(exe);
        } else {
            tokens.push_back(exe);
        }
        if (args.is_object() && args.contains("args") && args["args"].is_array())
            for (auto& a : args["args"])
                if (a.is_string()) tokens.push_back(a.get<std::string>());

        std::string cmdline;
        for (size_t i = 0; i < tokens.size(); i++) {
            if (i) cmdline.push_back(' ');
            cmdline += quoteArg(tokens[i]);
        }

        SECURITY_ATTRIBUTES sa{};
        sa.nLength = sizeof(sa);
        sa.bInheritHandle = TRUE;
        HANDLE inRead = nullptr, inWrite = nullptr, outRead = nullptr, outWrite = nullptr, errRead = nullptr, errWrite = nullptr;
        if (!CreatePipe(&inRead, &inWrite, &sa, 0) || !CreatePipe(&outRead, &outWrite, &sa, 0) ||
            !CreatePipe(&errRead, &errWrite, &sa, 0))
            throw EngawaError::io("CreatePipe failed");
        SetHandleInformation(inWrite, HANDLE_FLAG_INHERIT, 0);
        SetHandleInformation(outRead, HANDLE_FLAG_INHERIT, 0);
        SetHandleInformation(errRead, HANDLE_FLAG_INHERIT, 0);

        STARTUPINFOW si{};
        si.cb = sizeof(si);
        si.dwFlags = STARTF_USESTDHANDLES;
        si.hStdInput = inRead;
        si.hStdOutput = outWrite;
        si.hStdError = errWrite;
        PROCESS_INFORMATION pi{};
        std::wstring wcmd = ToWide(cmdline);
        std::vector<wchar_t> mutableCmd(wcmd.begin(), wcmd.end());
        mutableCmd.push_back(L'\0');
        std::wstring cwd = ToWide(bundleRoot_);

        // Spawn suspended so the process can be placed in a job (which owns its whole child tree)
        // BEFORE it runs — the sidecar image is often a launcher (cmd.exe → node) whose real work is a
        // child, and kill must reap the tree, not orphan node holding the stdout pipe open.
        BOOL ok = CreateProcessW(nullptr, mutableCmd.data(), nullptr, nullptr, TRUE,
                                 CREATE_NO_WINDOW | CREATE_SUSPENDED, nullptr, cwd.c_str(), &si, &pi);
        CloseHandle(inRead);
        CloseHandle(outWrite);
        CloseHandle(errWrite);
        if (!ok) {
            CloseHandle(inWrite);
            CloseHandle(outRead);
            CloseHandle(errRead);
            throw EngawaError::io("CreateProcess failed");
        }

        HANDLE job = CreateJobObjectW(nullptr, nullptr);
        if (job) {
            JOBOBJECT_EXTENDED_LIMIT_INFORMATION jeli{};
            jeli.BasicLimitInformation.LimitFlags = JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE;
            SetInformationJobObject(job, JobObjectExtendedLimitInformation, &jeli, sizeof(jeli));
            AssignProcessToJobObject(job, pi.hProcess);
        }
        ResumeThread(pi.hThread);
        CloseHandle(pi.hThread);

        int pid = static_cast<int>(pi.dwProcessId);
        auto st = std::make_unique<ProcState>();
        st->hProcess = pi.hProcess;
        st->job = job;
        st->pid = pi.dwProcessId;
        st->stdinWrite = inWrite;
        st->out.readEnd = outRead;
        st->err.readEnd = errRead;
        ProcState* raw = st.get();
        {
            std::lock_guard<std::mutex> lk(procsMu_);
            procs_[pid] = std::move(st);
        }

        startReader(raw, &raw->out, "stdout");
        startReader(raw, &raw->err, "stderr");
        std::thread([this, raw, pid] {
            WaitForSingleObject(raw->hProcess, INFINITE);
            DWORD code = 0;
            GetExitCodeProcess(raw->hProcess, &code);
            raw->exitCode = static_cast<int>(code);
            raw->exited = true;
            maybeExit(raw, pid);
        }).detach();

        return Json{{"pid", pid}};
    }

    // Windows-runnable extensions; the bare, extension-less declared path is skipped in favour of an
    // extension candidate (the repo ships both the Unix script and a .cmd).
    std::string resolveSidecar(const std::string& command, bool& isBatch) {
        isBatch = false;
        if (!sidecars_.count(command)) return "";  // allowlist is on the declared string
        std::string basePath = U8(fsys::absolute(P(bundleRoot_) / P(command)));

        std::vector<std::string> candidates;
        std::string declaredExt = lowerExt(command);
        bool declaredRunnable = false;
        for (auto* e : kExeExts)
            if (declaredExt == e) declaredRunnable = true;
        if (declaredRunnable) candidates.push_back(basePath);
        else
            for (auto* e : kExeExts) candidates.push_back(basePath + e);

        for (auto& cand : candidates) {
            if (!Files::isFile(cand)) continue;
            if (!Files::isInside(bundleRoot_, cand)) return "";  // §7.2 containment on the resolved path
            std::string ext = lowerExt(cand);
            isBatch = (ext == ".bat" || ext == ".cmd");
            return cand;
        }
        return "";
    }

    // ---- stdin / read / kill --------------------------------------------------------------
    Json stdinWrite(const Json& args) {
        // Validate data (EINVAL) before the pid lookup (ESRCH) so error precedence matches the mock host.
        std::string data = ja::reqStringAllowEmpty(args, "data");
        ProcState* st = require(args);
        if (st->exitEmitted) return Json(nullptr);  // already exited: no-op
        if (!st->stdinWrite) throw EngawaError::io("stdin is closed");  // after stdinClose
        DWORD written = 0;
        if (!data.empty() && !WriteFile(st->stdinWrite, data.data(), (DWORD)data.size(), &written, nullptr))
            throw EngawaError::io("stdin write failed");
        return Json(nullptr);
    }

    Json stdinClose(const Json& args) {
        ProcState* st = require(args);
        if (st->exitEmitted) return Json(nullptr);
        if (st->stdinWrite) { CloseHandle(st->stdinWrite); st->stdinWrite = nullptr; }
        return Json(nullptr);
    }

    Json read(const Json& args) {
        ProcState* st = require(args);
        auto streamArg = ja::optString(args, "stream");
        if (streamArg && *streamArg != "stdout" && *streamArg != "stderr")
            throw EngawaError::invalid("bad stream");
        int maxBytes = 65536;
        if (args.is_object() && args.contains("maxBytes") && !args["maxBytes"].is_null()) {
            if (!args["maxBytes"].is_number()) throw EngawaError::invalid("bad maxBytes");
            double mb = args["maxBytes"].get<double>();
            if (mb < 0) throw EngawaError::invalid("bad maxBytes");
            // mb >= 0 here, but still may exceed INT_MAX (e.g. 1e300); static_cast<int> of an
            // out-of-range double is UB, so clamp before the cast.
            if (mb > static_cast<double>(INT_MAX)) mb = static_cast<double>(INT_MAX);
            maxBytes = static_cast<int>(mb);
        }
        StreamBuf& which = (streamArg && *streamArg == "stderr") ? st->err : st->out;

        std::string data;
        bool eof;
        {
            std::lock_guard<std::mutex> lk(which.gate);
            int avail = static_cast<int>(which.size());
            int take = maxBytes < 0 ? 0 : (maxBytes > avail ? avail : maxBytes);
            int end = validUtf8End(which.head(), take);
            // Forward progress (§4.1): return a whole buffered character even if maxBytes was too small.
            if (end == 0 && !which.empty())
                end = validUtf8End(which.head(), avail < 4 ? avail : 4);
            data.assign(which.head(), static_cast<size_t>(end));
            which.consume(static_cast<size_t>(end));
            if (which.paused && !which.eof && which.size() < kCap) {
                which.paused = false;
                which.resume.notify_all();  // lift backpressure
            }
            eof = which.eof && which.empty();
        }
        maybeExit(st, st->pid);
        return Json{{"data", data}, {"eof", eof}};
    }

    Json kill(const Json& args) {
        ProcState* st = require(args);
        if (st->exitEmitted) return Json(nullptr);  // already exited: no-op, not ESRCH
        if (st->job) TerminateJobObject(st->job, 1);  // kill the whole tree (§ kill)
        else TerminateProcess(st->hProcess, 1);
        return Json(nullptr);
    }

    Json wasPaused(const Json& args) {
        ProcState* st = require(args);
        auto streamArg = ja::optString(args, "stream");
        StreamBuf& which = (streamArg && *streamArg == "stderr") ? st->err : st->out;
        std::lock_guard<std::mutex> lk(which.gate);
        return which.everPaused;
    }

    ProcState* require(const Json& args) {
        int pid = ja::reqInt(args, "pid");
        std::lock_guard<std::mutex> lk(procsMu_);
        auto it = procs_.find(pid);
        if (it != procs_.end()) return it->second.get();  // exited procs stay mapped (post-exit no-ops)
        throw EngawaError("ESRCH", "no such process: " + std::to_string(pid));
    }

    // ---- stream pump ----------------------------------------------------------------------
    void startReader(ProcState* st, StreamBuf* sb, const char* which) {
        std::string whichStr = which;
        std::thread([this, st, sb, whichStr] {
            std::vector<char> tmp(65536);
            for (;;) {
                {
                    std::unique_lock<std::mutex> lk(sb->gate);
                    sb->resume.wait(lk, [sb] { return !sb->paused; });  // block while paused (backpressure)
                }
                DWORD n = 0;
                if (!ReadFile(sb->readEnd, tmp.data(), (DWORD)tmp.size(), &n, nullptr) || n == 0) break;
                bool wasEmpty;
                {
                    std::lock_guard<std::mutex> lk(sb->gate);
                    wasEmpty = sb->empty();
                    sb->append(tmp.data(), n);
                    if (sb->size() >= kCap) { sb->paused = true; sb->everPaused = true; }
                }
                // §4.1: level-triggered readable when a drained stream goes non-empty; {pid,stream}
                // ARE the signal — no data bytes ride the event.
                if (wasEmpty) emitter_->emit("process.readable", Json{{"pid", (int)st->pid}, {"stream", whichStr}});
            }
            { std::lock_guard<std::mutex> lk(sb->gate); sb->eof = true; }
            maybeExit(st, (int)st->pid);
        }).detach();
    }

    // exit fires only after the process exited AND both streams are fully drained (§4.1).
    void maybeExit(ProcState* st, int pid) {
        if (!st->exited || st->exitEmitted) return;
        bool drained;
        {
            std::scoped_lock lk(st->out.gate, st->err.gate);
            drained = st->out.eof && st->err.eof && st->out.empty() && st->err.empty();
        }
        if (!drained) return;
        bool expected = false;
        if (!st->exitEmitted.compare_exchange_strong(expected, true)) return;  // emit exactly once
        // Reap handles: the process has exited, both streams are EOF + drained, and exitEmitted now
        // latches every other op (read/kill/stdin*) to a no-op — so nothing else will touch them. The
        // map entry stays as a drained tombstone (post-exit reads return eof, kill is a no-op). This is
        // the only place the process/job/pipe handles are closed, so each spawn is fully reclaimed.
        if (st->stdinWrite) { CloseHandle(st->stdinWrite); st->stdinWrite = nullptr; }
        if (st->out.readEnd) { CloseHandle(st->out.readEnd); st->out.readEnd = nullptr; }
        if (st->err.readEnd) { CloseHandle(st->err.readEnd); st->err.readEnd = nullptr; }
        if (st->job) { CloseHandle(st->job); st->job = nullptr; }
        if (st->hProcess) { CloseHandle(st->hProcess); st->hProcess = nullptr; }
        emitter_->emit("process.exit", Json{{"pid", pid}, {"code", st->exitCode}});
    }

    IEventEmitter* emitter_;
    bool conformance_;
    std::string bundleRoot_;
    std::unordered_set<std::string> sidecars_;
    std::mutex procsMu_;
    std::unordered_map<int, std::unique_ptr<ProcState>> procs_;
};

}  // namespace

std::unique_ptr<IAdapter> makeProcessAdapter(IEventEmitter* emitter, const HostOptions& opts) {
    return std::make_unique<ProcessAdapter>(emitter, opts);
}
